### PR TITLE
openvpn: Update to 2.4.6, cleanup

### PR DESCRIPTION
**Pull request by diizzyy**
This updates OpenVPN to 2.4.6 and does a general cleanup

Signed-off-by: Daniel Engberg <daniel.engberg.lists@pyret.net>

### CODE REVIEW

**Comment by diizzyy** [comment]
@dedeckeh @mkrkn @hauke @neheb @pprindeville

File: package/network/services/openvpn/files/openvpn.config [hunk 1]
```diff
@@ -398,7 +398,7 @@ config openvpn sample_client
 	# LZ4 requires OpenVPN 2.4+ on server and client
 #	option compress lz4
 	# LZO is compatible with most OpenVPN versions
-	option compress lzo
+#	option compress lzo
```
Comment (pprindeville) [inline package/network/services/openvpn/files/openvpn.config]: So we don't have ANY compression modes?  Why is that?

Comment (diizzyy) [reply package/network/services/openvpn/files/openvpn.config]: https://serverfault.com/a/755690
https://arstechnica.com/civis/viewtopic.php?p=1307910#p1307910
https://www.snbforums.com/threads/openvpn-server-settings-connection-very-slow.37998/#post-313327
etc
Pretty much similar to my experience on various platforms.

**Comment by neheb** [comment]
I don't use OpenVPN. Otherwise the changes look good.

Compression is pointless with TLS. I find it funny that the first link complained about performance on an x86 CPU when most people are using MIPS...

**Comment by diizzyy** [comment]
Compile tested: mvebu, Linksys WRT3200ACM, OpenWrt master
Runtime tested: mvebu, Linksys WRT3200ACM, OpenWrt master

File: package/network/services/openvpn/Makefile [hunk 2]
```diff
@@ -73,34 +72,30 @@ CONFIGURE_VARS += \
...
-		$(if $(CONFIG_OPENVPN_$(BUILD_VARIANT)_ENABLE_PF),--enable,--disable)-pf \
-		$(if $(CONFIG_OPENVPN_NOSSL),--disable-ssl --disable-crypto,--enable-ssl --enable-crypto) \
-		$(if $(CONFIG_OPENVPN_OPENSSL),--with-crypto-library=openssl) \
-		$(if $(CONFIG_OPENVPN_MBEDTLS),--with-crypto-library=mbedtls) \
-	)
-endef
+TARGET_CFLAGS += -ffunction-sections -fdata-sections -flto
+TARGET_LDFLAGS += -Wl,--gc-sections -flto
+
+CONFIGURE_ARGS += \
```
Comment (pprindeville) [inline package/network/services/openvpn/Makefile]: Can you put the flags back into the same order as before so it's easier to tell if what's changed?  Thanks.

Comment (diizzyy) [reply package/network/services/openvpn/Makefile]: I made this change as the old one was just a mash making it hard to keep track of what's been changed upstream.
SOCKS, HTTP and --enable/disable-ssl are removed in upstream (before 2.4.5), x509-alt-username doesn't have -ssl at the end. PORT_SHARE option most likely got lost for this very reason.

File: package/network/services/openvpn/Makefile [hunk 3]
```diff
@@ -73,34 +72,30 @@ CONFIGURE_VARS += \
...
+	--disable-plugins \
+	--disable-pkcs11 \
+	--disable-debug \
+	--disable-selinux \
+	--disable-systemd \
+	$(if $(CONFIG_OPENVPN_$(BUILD_VARIANT)_ENABLE_LZO),--enable,--disable)-lzo \
+	$(if $(CONFIG_OPENVPN_$(BUILD_VARIANT)_ENABLE_LZ4),--enable,--disable)-lz4 \
+	$(if $(CONFIG_OPENVPN_NOSSL),--disable-crypto,--enable-crypto) \
+	$(if $(CONFIG_OPENVPN_$(BUILD_VARIANT)_ENABLE_X509_ALT_USERNAME),--enable,--disable)-x509-alt-username \
```
Comment (pprindeville) [inline package/network/services/openvpn/Makefile]: Did you mean to change from:

```
		$(if $(CONFIG_OPENVPN_$(BUILD_VARIANT)_ENABLE_X509_ALT_USERNAME),enable,disable-x509-alt-username)-ssl \
```

to:

```
	$(if $(CONFIG_OPENVPN_$(BUILD_VARIANT)_ENABLE_X509_ALT_USERNAME),--enable,--disable)-x509-alt-username \
```

File: package/network/services/openvpn/Makefile [hunk 4]
```diff
@@ -73,34 +72,30 @@ CONFIGURE_VARS += \
...
-		--disable-systemd \
-		--disable-plugins \
-		--disable-debug \
-		--disable-pkcs11 \
-		$(if $(CONFIG_OPENVPN_$(BUILD_VARIANT)_ENABLE_LZO),--enable,--disable)-lzo \
-		$(if $(CONFIG_OPENVPN_$(BUILD_VARIANT)_ENABLE_LZ4),--enable,--disable)-lz4 \
-		$(if $(CONFIG_OPENVPN_$(BUILD_VARIANT)_ENABLE_X509_ALT_USERNAME),enable,disable-x509-alt-username)-ssl \
-		$(if $(CONFIG_OPENVPN_$(BUILD_VARIANT)_ENABLE_SERVER),--enable,--disable)-server \
-		$(if $(CONFIG_OPENVPN_$(BUILD_VARIANT)_ENABLE_MANAGEMENT),--enable,--disable)-management \
-		$(if $(CONFIG_OPENVPN_$(BUILD_VARIANT)_ENABLE_SOCKS),--enable,--disable)-socks \
```
Comment (pprindeville) [inline package/network/services/openvpn/Makefile]: This line is being dropped...

File: package/network/services/openvpn/Makefile [hunk 5]
```diff
@@ -73,34 +72,30 @@ CONFIGURE_VARS += \
...
-		$(if $(CONFIG_OPENVPN_$(BUILD_VARIANT)_ENABLE_HTTP),--enable,--disable)-http-proxy \
```
Comment (pprindeville) [inline package/network/services/openvpn/Makefile]: And this one...

File: package/network/services/openvpn/Makefile [hunk 6]
```diff
@@ -73,34 +72,30 @@ CONFIGURE_VARS += \
...
-		$(if $(CONFIG_OPENVPN_$(BUILD_VARIANT)_ENABLE_FRAGMENT),--enable,--disable)-fragment \
-		$(if $(CONFIG_OPENVPN_$(BUILD_VARIANT)_ENABLE_MULTIHOME),--enable,--disable)-multihome \
-		$(if $(CONFIG_OPENVPN_$(BUILD_VARIANT)_ENABLE_IPROUTE2),--enable,--disable)-iproute2 \
-		$(if $(CONFIG_OPENVPN_$(BUILD_VARIANT)_ENABLE_DEF_AUTH),--enable,--disable)-def-auth \
-		$(if $(CONFIG_OPENVPN_$(BUILD_VARIANT)_ENABLE_PF),--enable,--disable)-pf \
-		$(if $(CONFIG_OPENVPN_NOSSL),--disable-ssl --disable-crypto,--enable-ssl --enable-crypto) \
```
Comment (pprindeville) [inline package/network/services/openvpn/Makefile]: And this one.

File: package/network/services/openvpn/Makefile [hunk 7]
```diff
@@ -73,34 +72,30 @@ CONFIGURE_VARS += \
...
+	--disable-systemd \
+	$(if $(CONFIG_OPENVPN_$(BUILD_VARIANT)_ENABLE_LZO),--enable,--disable)-lzo \
+	$(if $(CONFIG_OPENVPN_$(BUILD_VARIANT)_ENABLE_LZ4),--enable,--disable)-lz4 \
+	$(if $(CONFIG_OPENVPN_NOSSL),--disable-crypto,--enable-crypto) \
+	$(if $(CONFIG_OPENVPN_$(BUILD_VARIANT)_ENABLE_X509_ALT_USERNAME),--enable,--disable)-x509-alt-username \
+	$(if $(CONFIG_OPENVPN_$(BUILD_VARIANT)_ENABLE_SERVER),--enable,--disable)-server \
+	$(if $(CONFIG_OPENVPN_$(BUILD_VARIANT)_ENABLE_MANAGEMENT),--enable,--disable)-management \
+	$(if $(CONFIG_OPENVPN_$(BUILD_VARIANT)_ENABLE_FRAGMENT),--enable,--disable)-fragment \
+	$(if $(CONFIG_OPENVPN_$(BUILD_VARIANT)_ENABLE_MULTIHOME),--enable,--disable)-multihome \
+        $(if $(CONFIG_OPENVPN_$(BUILD_VARIANT)_ENABLE_PORT_SHARE),--enable,--disable)-port-share \
```
Comment (pprindeville) [inline package/network/services/openvpn/Makefile]: Fix indent.

File: package/network/services/openvpn/Makefile (hunk 3 above)
Comment (diizzyy) [reply package/network/services/openvpn/Makefile]: Yes, see below
```
./configure --help
...
  --enable-x509-alt-username
                          enable the --x509-username-field feature
                          [default=no]
```
As a side note, default values (that configure lists) gets ignored for whatever reason hence all the disable/enable options.

File: package/network/services/openvpn/Makefile (hunk 5 above)
Comment (diizzyy) [reply package/network/services/openvpn/Makefile]: See previous comment and/or commit msg

File: package/network/services/openvpn/Makefile (hunk 4 above)
Comment (diizzyy) [reply package/network/services/openvpn/Makefile]: See previous comment and/or commit msg

File: package/network/services/openvpn/Makefile (hunk 6 above)
Comment (diizzyy) [reply package/network/services/openvpn/Makefile]: See previous comment and/or commit msg

File: package/network/services/openvpn/Makefile (hunk 7 above)
Comment (diizzyy) [reply package/network/services/openvpn/Makefile]: Sigh, is there an easy way to do this or do I need to revert to previous commit, rebase and apply the backed out commit again?

Comment (pprindeville) [reply package/network/services/openvpn/Makefile]: um...  I do this...  If it's 5 commits back, then:

```
% git rebase -i HEAD~5
```

change the `pick` to `edit` on the commit you want to fix.  Write/quit from editor.

When it stops, do:

```
% vi file-to-change
```

make your change, then:

```
% git add file-to-change
% git commit --amend
```

write/quit with no changes

```
% git rebase --continue
```

File: package/network/services/openvpn/Makefile (hunk 5 above)
Comment (diizzyy) [reply package/network/services/openvpn/Makefile]: Fixed

File: package/network/services/openvpn/Makefile (hunk 7 above)
Comment (diizzyy) [reply package/network/services/openvpn/Makefile]: Fixed

**Comment by jow-** [comment]
Is there any way to get any of the disabled features back in form of an installable binary package? It seems to me that you disabled them everywhere, including the openssl variant which I would've expected to be the `full` one.

**Comment by jow-** [comment]
What is the build time impact of `-flto` ?

**Comment by diizzyy** [comment]
No, I'm not sure why you'd expect openssl to be a full package in the first place (no build has never been packed with all features). 
```-flto``` add a few seconds tops on a G3220, it doesn't slow down compilation by X times.

**Comment by jow-** [comment]
> No, I'm not sure why you'd expect openssl to be a full package in the first place

I do not expect that, but I do not expect already available features to suddenly vanish without replacement either.

**Comment by diizzyy** [comment]
There's no replacement because of the simple reason that there's none. Compression in OpenVPN has a huge performance penalty with very minor advantages if any especially on embedded systems. Given that size is often used as a reason I assumed that this wouldn't be an issue especially when it essentially doesn't bring any advantages. Are you suggesting that we should leave it in because of POLA (https://www.freebsd.org/doc/en_US.ISO8859-1/books/handbook/freebsd-glossary.html#pola-glossary)?
You can of course compile your own package with it enabled if you're dead set on using it.

Unless I misunderstand the documentation and source code of OpenVPN it's applied per packet (if applied) which isn't very effective as the chunk/block size is small.

**Comment by jow-** [comment]
This update will effectively break existing configurations, I just checked. In my case an opkg upgrade of the package would kill two of the three tunnels I have running here.

It might be that LZO compression is ineffective or a security weakness, and we can even argue about educating users about not using it but as long as any `comp-*` token is present in actively used OpenVPN configurations, it should keep working using opkg installable binary packages.

In reality, a large chunk of configurations is likely copy pasted from random sources or has been handed over by employers or commercial service providers. If these configs happen to enable LZO compression, they should continue to work.

Really, the burden of these micro optimizations should be put on those users doing custom builds anyway, not on those being dependent on binary package distributions.

**Comment by jow-** [comment]
It seems that at the very least you should use `--enable-lzo-stub` to keep interoperability with existing configs. But that is something you should've researched yourself before proposing to disable feature support by default.

**Comment by diizzyy** [comment]
From what I read it's negotiable but apparently that didn't seem to be completely accurate, I apologize.
If the server doesn't push settings to enable compression (which most providers seems to do, PIA, AirVPN etc) it works just fine however if --compress is added to the server config it enables "packet framing for compression" which causes clients that doesn't support compression to stop working. In short, it enforces instead of negotiating. :-/

I'll add back lzo compression...

**Comment by diizzyy** [comment]
Backed out removal of compression support by default due to request

**Comment by pachulo** [comment]
It seems like OpenVPN 2.4.6 is a security update: https://github.com/OpenVPN/openvpn/blob/release/2.4/Changes.rst#bug-fixes
Can we move the update to another PR and leave here the cleanup?

**Comment by diizzyy** [comment]
The CVE is for Windows...
The cleanup is necessary as it fixes several (two) menu options.

**Comment by diizzyy** [comment]
@jow- 
Anything else than should be changed?

**Comment by diizzyy** [comment]
@mkrkn

**Comment by mkrkn** [comment]
To remove any doubt, an OpenVPN peer with no "compress" option set cannot talk to a peer with any "compress" option set. They will be able to negotiate TLS for the control channel, but will be unable to establish a data channel.

I think @jow- is right in that a lot of configs are copy-pasted. An issue on top of that is that OpenVPN doesn't notify the user that mismatching compression options are causing an issue. OpenVPN will negotiate TLS, but will not be able to pass traffic beyond that, and will simply behave as if the other side is not responding. 

As a non-breaking educational approach we could log a warning from the init script if any compression options (compress or comp-lzo) are set.

Other than that I don't think I have anything to add to this PR, it looks good to me.

**Comment by diizzyy** [comment]
Anything else that needs to be done?

**Comment by diizzyy** [comment]
@jow- 
Ok to commit?

**Comment by neheb** [comment]
Does not apply.

May wish to mention that this fixes CVE-2018-7544

**Comment by diizzyy** [comment]
I'll update during the weekend

**Comment by diizzyy** [comment]
@hauke 
Can you have a look?

**Comment by diizzyy** [comment]
Compile/Runtime tested on: ramips, D-Link DIR-860L B1, OpenWrt master

**Comment by diizzyy** [comment]
@dedeckeh 
Can we merge this now? It doesn't change any of the default settings

**Comment by dedeckeh** [comment]
openvpn-mbedtls does not seem to support x509-alt-username as I get the following build error :

> configure: error: mbed TLS does not support the --x509-username-field feature
make[2]: *** [Makefile:135: /home/dedeckeh/arm-virt-staging/build_dir/target-arm_cortex-a15+neon-vfpv4_musl_eabi/openvpn-mbedtls/openvpn-2.4.6/.configured_68b329da9893e34099c7d8ad5cb9c940] Error 1

**Comment by diizzyy** [comment]
@dedeckeh 
Fixed, thanks

**Comment by jow-** [comment]
I decided to push an alternative change that merely updates the package and fixes the existing options. Please open a separate PR for your LTO and ordering improvements. Also why did you remove the autoconf fixup?

**Comment by diizzyy** [comment]
After several months, convenient

**Comment by EXSERENS** [comment]
hahahahahaha finally it's merged

**Pull request closed without merging**